### PR TITLE
Fix #15858 - Long compiler identification in ELF ##bin

### LINKED ISSUE #15858: ELF `.comment` section parsing - compiler name
I have a file (can't share) with the following content in the `.comment` section:
image

As you can see it contains many strings of the used compiler. Question is - what is the best course of action in this case? To parse this mess - cut the rest?
This is what `i` outputs:
```
0x0000262c]> i
fd       3
size     0x5e4ec
humansz  377.2K
mode     r-x
format   elf
iorw     false
blksz    0x0
block    0x100
type     EXEC (Executable file)
arch     ppc
baddr    0x0
binsz    385882
bintype  elf
bits     32
canary   false
class    ELF32
compiler /GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)//GCC: (GNU) 3.3.1 (qnx-nto)
crypto   false
endian   big
havecode true
laddr    0x0
lang     c
linenum  false
lsyms    false
machine  PowerPC
maxopsz  4
minopsz  4
nx       false
os       linux
pcalign  4
pic      false
relocs   false
rpath    NONE
sanitiz  false
static   true
stripped true
subsys   linux
va       true

[0x0000262c]> 
```

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

**Comment by XVilka** [comment]
I confirm - it works. The only test that probably should be fixed is:
```
[XX] db/formats/elf/elf-riscv64 ELF: riscv64: file format  278
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i /tmp/tmp-171051B09xjkkd3Jb ../bins/elf/analysis/guess-number-riscv64
if
 fd       3
 file     ../bins/elf/analysis/guess-number-riscv64
 size     0x9d5a8
 humansz  629.4K
 mode     r-x
 format   elf64
 iorw     false
 blksz    0x0
 block    0x100
 type     EXEC (Executable file)
 arch     riscv
 baddr    0x10000
 binsz    643108
 bintype  elf
 bits     64
 canary   false
 class    ELF64
-compiler GCC: (GNU) 5.2.0
+compiler GCC: (GNU) 5.2.0 
```

**Comment by radare** [comment]
It needs more work filtering bad chars and suchZ i prefer to wait for 4.3 for a proper fix

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -3967,20 +3967,10 @@ char *Elf_(r_bin_elf_compiler)(ELFOBJ *bin) {
 		free (buf);
 		return NULL;
 	}
-
+	char *nullbyte = buf + strlen (buf);
+	if (!nullbyte[0] && nullbyte[1]) {
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c]: !nullbyte[0] Is always true, I think you can just keep the second part

File: libr/bin/format/elf/elf.c [hunk 2]
```diff
@@ -3967,20 +3966,11 @@ char *Elf_(r_bin_elf_compiler)(ELFOBJ *bin) {
 		free (buf);
 		return NULL;
 	}
-
+	const size_t buflen = strlen (buf);
+	char *nullbyte = buf + buflen;
+	if (nullbyte[1] && buflen < sz) {
+		nullbyte[0] = ' ';
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c]: Shall we keep using `/`? There would be less tests to fix and it will be more consistent with current behaviour.

Comment (ret2libc) [reply libr/bin/format/elf/elf.c]: Also, why are we doing this? Didn't you say in https://github.com/radareorg/radare2/issues/15858#issuecomment-576242135 that we should stop at the first null byte?

**Pull request merged**